### PR TITLE
fix skybox nesting

### PR DESCRIPTION
**Pull request by zzmp**

### CODE REVIEW

**Comment by zzmp** [comment]
QA:

Test this for skyboxes and procedural skyboxes:

Make sure the old behavior still works - do this with only a single skybox.
- If there is a procedural skybox set, it should be shown, not the default (it may need to load first)
- If there is no skybox set, the default horizon should show
- If there is a color set, the color should be used as the skybox (this is fixed to include 0,0,0)
- If there is a texture set, the texture should be used as the skybox
- If there is a color and a texture set, a tinted version of the texture should be used as the skybox (this is fixed to include 0,0,0)
- When you turn off your skybox, the default should come back

Now test the new behavior, nested skyboxes! The only thing that nests is the actual skybox color and/or texture.
- If there are multiple zones, the smallest (volumed) zone should give the lighting
  - This includes ambient, with the exception that if ambient texture is unset, it will be taken from the skybox
- If this inner (smallest) zone has a skybox, it should be used
- If it does not, the next layered zone (i.e. next smallest) will be checked for a skybox, etc.
- If not zones that contain you have a skybox, the default should be used
- Note that "has a skybox" means that the skybox is set to something, whether it is a color or a texture or both
- As before, moving around should update skyboxes appropriately
- As before, changing your zones should update what you see immediately
  - Try adding/removing the skybox from your inner zone, and watch the next one appear
  - Try having a few layers of zone with no skybox, and setting arbitrary zone's skybox

**Comment by hifi-gustavo** [comment]
Build console output for commit: d45f5e92844677a3e0e7947af82d275df03dc2fd available at: https://gist.github.com/e7aa0bd3e833113c6a699223fde06ecd<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8530-d45f5e9.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8530-d45f5e9.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 841e341e69eb10f64754d103818cfde9faa13379 available at: https://gist.github.com/e7aa0bd3e833113c6a699223fde06ecd<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8530-841e341.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8530-841e341.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: dcfbc145dced64a56a61b358790cb17b3a8675d3 available at: https://gist.github.com/e7aa0bd3e833113c6a699223fde06ecd<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8530-dcfbc14.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8530-dcfbc14.dmg

**Comment by cozza13** [comment]
ok to test

**Comment by hifi-gustavo** [comment]
Build console output for commit: dcfbc145dced64a56a61b358790cb17b3a8675d3 available at: https://gist.github.com/e7aa0bd3e833113c6a699223fde06ecd<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8530-dcfbc14.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8530-dcfbc14.dmg

**Comment by Nex-Pro** [comment]
Testing on Win10 in progress.

**Comment by Nex-Pro** [comment]
**PR #8530**

**Testing Architecture:** Desktop PC
**CPU:** 4.4 ghz Intel i7
**RAM:** 16 GB
**GPU:** Nvidia gtx 1070 8gb

**Testing OS:** win 10 pro

**QA FAIL** On my system.


**Starting condition:** 
    Ensured no previous install of this PR.
    Ran Dev build and requested a reset of settings.
    Proceeded with installation.
    Selected not to copy content ensuring fresh content at localhost.
    Launched this build of interface by using the 'go home'.
    Searched entity list for "zone" to ensure there were no zones that I was unaware of.


**TESTS:**

If there is a procedural skybox set, it should be shown, not the default (it may need to load first): **SUCCESS**
	
.        Obtained day/night cycle from market. It started immediately.
.        Restarted interface, day/night cycle was showing after no more than 2 seconds.	



If there is no skybox set, the default horizon should show:

.     ***This is slightly ambiguous therefore I tested for:***

.        **With the default skybox:**

.           "Nothing" selected in the BACKGROUND dropdown of the default skybox.: **Default horizon was displayed**

.           Reselect "Skybox" in the  BACKGROUND dropdown of the default skybox.: **Sky turned black.**
.           Stayed black for some time despite atp:/SKY-Home-2-Lights-8k-2.jpg showing in the Skybox URL.
.           Swapping back to Nothing displayed the default horizon then reselecting Skybox resulted in the sky .           turning white.
.           Several right click drags on the screen to change view angle provoked the sky to turn black.
.           Closing and restarting  interface made the default skybox re-appear.

        **With no skyboxes:**

.          Black sky.        


If there is a color set, the color should be used as the skybox (this is fixed to include 0,0,0): 
.      **Setting  0, 0, 0 has no effect.** I feel this should display as pure black.
.      Other colors work to tint the sky.

.      **There are discreprancies between the colors selected for the sky and the colours displayed.**

.        Inspecting screengrabs that include the color picker and resulting sky colour reveals....
 
.         Pure red value in the picker reads correctly as 255, 0, 0  picking from the sky shows **252**, 0, 0
.
.         Pure green value in the picker reads correctly as 0, 255, 0  picking from the sky shows 0, **252**, 0
.
.         Pure blue value in the picker reads correctly as 255, 0, 0  picking from the sky shows 0, 0, **253**
.
.         Small differences of around 1% but inconsistent.



If there is a texture set, the texture should be used as the skybox: **FAIL**

.        Only works after restarting interface. But does not always work. Tried with a dropbox link to no avail.
.        Problem was persistent over many attempts.



If there is a color and a texture set, a tinted version of the texture should be used as the skybox (this is fixed to include 0,0,0):


When you turn off your skybox, the default should come back: Does this mean turn off the Skybox setting in the zone or removing any zones ? I checked both conditions.

.       When I set the zone to have 'Nothing' I get the default blue horizon.
.       When I remove all zones I get a black sky.

**Additional observations:**

Zone entity with no skybox or textures has sky-blue upper polygon in edit mode. ( visual cue ?) This is what is producing the blue sky we occasionally see. I set the width of a xone to be 10 and the length to be 30000 and I got a line across the sky. ( see image below ).


**Feedback**:

Setting 'Nothing' in a skybox shows the default horizon. Surely instead of Nothing this should be called "Default" or some other appropriate name as it is clearly not 'nothing'.

Why is there no setting to change the colour of the ambient? I can load an image of a solid colour to achieve this. Why not have the option to dial in a colour when there is no image? 

It would be nice to have the option to change keylight colour in Kelvins. Same for solid ambient.

***See the second chart on this page: https://en.wikipedia.org/wiki/Color_temperature**

I would like to make a PR to re-arrange the skybox entity properties so we are not forced to scroll when whe use the skybox background dropdown or the colour picker.

Why is the collision shape for a skybox not set to 'None'?

no_zone
no_zone
pr8530_noskyboxtexture_showing


@zzmp  let me know if you want me to test the nested skyboxes or hold-off due to what I have seen on the first part of the QA.

**Comment by jherico** [comment]
When entering Dreaming, I see the main skybox, but if I exit the zone, I don't see the default skybox.  When I re-enter the zone, the procedural skybox does not re-appear.

**Comment by zzmp** [comment]
Split allow black skybox to its own PR: #8535.

**Comment by zzmp** [comment]
ok @Nex-Pro ready for another go.
I'm not sure what that blue polygon is. Was it there before?
I wouldn't worry about color values being slightly off - we do some tonemapping that may account for that.

**Comment by hifi-gustavo** [comment]
Build console output for commit: f30308b68c81bee240767ba79de8816deee280c4 available at: https://gist.github.com/e7aa0bd3e833113c6a699223fde06ecd<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8530-f30308b.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8530-f30308b.dmg

**Comment by Nex-Pro** [comment]
@zzmp I don't think the blue polygon is specific to this PR.

**Comment by Nex-Pro** [comment]
@zzmp 

Single skybox tests:

If there is a procedural skybox set, it should be shown, not the default (it may need to load first): **SUCCESS**


If there is no skybox set, the default horizon should show: **SUCCESS**


If there is a color set, the color should be used as the skybox (this is fixed to include 0,0,0): **SUCCESS**


If there is a texture set, the texture should be used as the skybox: **FAIL**
    Texture will not show first time it is set. Even if it is the default home scene ATP link. texture.
    I need to restart interface for any skybox texture to show.
    After restarting interface the skybox texture shows and persists over multiple changes from Skybox to Nothing. ( as long as I don't change the texture)
    Is this somehow a cacheing issue ?


If there is a color and a texture set, a tinted version of the texture should be used as the skybox (this is 
fixed to include 0,0,0): I can't decide what "fixed to include 0,0,0" Means. If I set a colour it tints the image. Why does 0,0,0 have the same effect as 255, 255, 255 ?  When no texture is set the sky is black at 0, 0 , 0. When a texture is set the sky looks the same at 255, 255, 255 and 0, 0, 0.


When you turn off your skybox, the default should come back: **SUCCESS**

**Comment by zzmp** [comment]
@Nex-Pro this didn't happen on my own tests...I'll have to retest tomorrow and see if anything different is happening.

**Comment by zzmp** [comment]
Retested, good find. Fixing it up right now.

**Comment by zzmp** [comment]
Dear Code Reviewer,

I thought that by rolling my own data structure, inherited from a std::set, I could abstract and simplify this feature. Having implemented it, I think I was wrong. Most of my problems in implementation came from forgetting about things like copying private members (`_skyboxLayer`, in particular) - things that the std::set made seem more simple.

If you have any better strategies, any method renamings, or any alternate structures that you think could have helped, please let me know.

Zach

**Comment by hifi-gustavo** [comment]
Build console output for commit: 0e9ff04c4ca6834afadcfe9f7cce25148876aa40 available at: https://gist.github.com/e7aa0bd3e833113c6a699223fde06ecd<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8530-0e9ff04.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8530-0e9ff04.dmg

**Comment by cozza13** [comment]
I tested this and it works. QA pass

File: libraries/entities-renderer/src/EntityTreeRenderer.cpp [hunk 1]
```diff
@@ -371,7 +362,7 @@ void EntityTreeRenderer::applyZonePropertiesToScene(std::shared_ptr<ZoneEntityIt
         sceneKeyLight->setAmbientMap(nullptr);
 
         skyStage->setBackgroundMode(model::SunSkyStage::SKY_DEFAULT);
-        return;
+        return false;
     }
```
Comment (AndrewMeadows) [inline libraries/entities-renderer/src/EntityTreeRenderer.cpp]: `EntityTreeRenderer::applyZoneAndHasSkybox()` only ever returns `false`.  Did you forget a return path somewhere?  If not then why return a bool value?  Are you planning on making it work correctly someday?

File: libraries/entities-renderer/src/EntityTreeRenderer.h [hunk 2]
```diff
@@ -185,26 +182,62 @@ public slots:
...
+        LayeredZones& operator=(const LayeredZones&) = delete;
+        LayeredZones& operator=(LayeredZones&&) = delete;
 
+        void clear();
+        std::pair<iterator, bool> insert(const LayeredZone& layer);
+
+        void apply();
+        void update(std::shared_ptr<ZoneEntityItem> zone);
+
+        bool contains(const LayeredZones& other);
```
Comment (AndrewMeadows) [inline libraries/entities-renderer/src/EntityTreeRenderer.h]: Overrides should be labeled as such.  We have a new warning about this enabled on linux.

**Pull request merged**